feat: merkle tree persistence

richard-ramos opened the pull request:
# Description
Persist Merkle Tree in sled db
This is part of Milestone: https://github.com/waku-org/go-waku/issues/605

# Changes

- Resume onchain sync from persisted tree db
- Close eth client and DB connection appropriately
- Pass in the path to the tree db

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_multiplication_x: | f61b4100 | [#1](https://ci.status.im/job/go-waku/job/prs/job/nix-flake/job/PR-630/1/) | 2023-08-07 21:33:15 | ~29 sec | `nix-flake` | [:page_facing_up:`log`](https://ci.status.im/job/go-waku/job/prs/job/nix-flake/job/PR-630/1/consoleText) |

richard-ramos commented:
:thinking: For some reason, when a tree config is passed in https://github.com/waku-org/go-waku/pull/630/files#diff-f9f1366bf48973e91af5a769c566b2662c33177e31388325c7f9307b4e7cca55R72 RLN refuses to initialize. Tomorrow I'll figure out what's going on here so I can properly test interop between nwaku and go-waku

richard-ramos commented:
The issue was that I was using the same database for all nodes in the test units, so RLN was complaining about not being able to acquire locks on the DB

chaitanyaprem commented on waku/v2/protocol/rln/group_manager/dynamic/metadata.go:
```diff
@@ -0,0 +1,39 @@
+	"errors"
+)
+
+type RLNMetadata struct {
```
Need to add comments for all the exported types.

chaitanyaprem commented on waku/v2/protocol/rln/group_manager/dynamic/dynamic.go:
```diff
@@ -278,17 +297,28 @@ func (gm *DynamicGroupManager) persistCredentials() error {
+
+		// TODO: should we track indexes to identify missing?
+		startIndex := rln.MembershipIndex(uint(oldestIndexInBlock.Int64()))
+		err := gm.rln.InsertMembers(startIndex, idCommitments)
```
Does the RLN API take care of a scenario where few members are already inserted and not throw error? If not, this has to be handled at this level.

e.g: While inserting members, let's say member insertion succeeded whereas before metadata got updated, node went down. Next time when node comes up, it will try to insert some set of members which are already inserted and probably a new set if they are part of a later block. In this case, atomicOp should not fail. If it fails, then we may have to handle it at this layer(identify if members are already present and not insert them again).

richard-ramos replied:
> Does the RLN API take care of a scenario where few members are already inserted and not throw error? I

The RLN API handles it by acting as an upsert, overwriting whatever is stored at some index in the merkle tree, so atomicOp should not fail as it will just replace the same members at the same position with the same values

richard-ramos commented on waku/v2/protocol/rln/group_manager/dynamic/web3.go:
```diff
@@ -112,7 +112,17 @@ func (gm *DynamicGroupManager) HandleGroupUpdates(ctx context.Context, handler R
 
 func (gm *DynamicGroupManager) loadOldEvents(ctx context.Context, rlnContract *contracts.RLN, handler RegistrationEventHandler) error {
-	events, err := gm.getEvents(ctx, 0, nil)
+
```
It should be possible to retrieve events with a full node because these type of nodes contain all transaction receipts/logs. A full node can still be costly so I agree that we'll need to come up with a solution for obtaining the merkle tree changes, instead of relying only on the event logs (although for the time being, it should be fine with using infura/pocket)

richard-ramos commented:
@chaitanyaprem I updated this PR with fixes for code review observations, most code climate issues, and also fixed some interop issues I found while attempting to get messages going back and forth between nwaku's chat2 and go-waku's

chaitanyaprem commented on cmd/waku/flags_rln.go:
```diff
@@ -30,7 +31,7 @@ func rlnFlags() []cli.Flag {
 		&cli.StringFlag{
 			Name:        "rln-relay-content-topic",
-			Value:       "/toy-chat/2/luzhou/proto",
+			Value:       protocol.NewContentTopic("toy-chat", 3, "mingde", "proto").String(),
```
Does it make sense to have a default value for content-topic? Rather we should throw an error if content-topic is not mentioned. 
We can always inlcude in docs which content-topic can be used if someone wants to just test.

alrevuelta replied:
Note that we plan to remove rln pubsub topics and content topics, so that rln is applied in all the subscribed ones if enabled.

See 1. from https://github.com/waku-org/nwaku/issues/1906
and https://discord.com/channels/1110799176264056863/1111540684575477821/1141378004841410640

richard-ramos replied:
For this PR let's keep it with the hardcoded default content topic to match nwaku's flag: https://github.com/waku-org/nwaku/blob/master/apps/wakunode2/external_config.nim#L165

chaitanyaprem commented on examples/chat2/chat.go:
```diff
@@ -77,16 +82,19 @@ func NewChat(ctx context.Context, node *node.WakuNode, connNotifier <-chan node.
+						chat.C <- e
+					}
+				}()
+			}
```
Shouldn't the wait-group be based on number of times we loop?
It is hardcoded to 7 in line 101.

richard-ramos replied:
That 7 refers to the goroutines created from line 102 to 113:
```
go chat.parseInput()
go chat.receiveMessages()
go chat.welcomeMessage()
go chat.connectionWatcher(&connectionWg, connNotifier)
go chat.staticNodes(&connectionWg)
go chat.discoverNodes(&connectionWg)
go chat.retrieveHistory(&connectionWg)
```

chaitanyaprem replied:
Ah..got it.

chaitanyaprem reviewed (approved):
LGTM other than few minor comments.
